### PR TITLE
On turn error on Bot Adapter

### PR DESCRIPTION
**Pull request by axelsrz**
- Added on_turn_error functionality to bot_adapter with updates to all related samples 
- Ported bot assert
- Added some missing annotations.

*This PR address issue #167

### CODE REVIEW

File: libraries/botbuilder-core/botbuilder/core/bot_adapter.py [hunk 1]
```diff
@@ -48,12 +50,25 @@ def use(self, middleware):
         """
         self._middleware.use(middleware)
 
-    async def run_middleware(self, context: TurnContext, callback: Callable=None):
+    async def run_middleware(self, context: TurnContext, callback: Callable[[TurnContext], Awaitable]= None):
```
Comment (cleemullins) [inline libraries/botbuilder-core/botbuilder/core/bot_adapter.py]: I don't understand the drift here from the C# codebase. The C# has [RunPipelineAsync](https://github.com/microsoft/botbuilder-dotnet/blob/ccf1643e3168a395f327d9eb3b6c351fa2855493/libraries/Microsoft.Bot.Builder/BotAdapter.cs#L158). 

I'm not clear why the name here is different, as it's going to cause a documentation confusion.

File: libraries/botbuilder-core/botbuilder/core/bot_assert.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
+from typing import List
```
Comment (cleemullins) [inline libraries/botbuilder-core/botbuilder/core/bot_assert.py]: Copyright / License statement is missing.

File: libraries/botbuilder-core/botbuilder/core/bot_adapter.py (hunk 1 above)
Comment (axelsrz) [reply libraries/botbuilder-core/botbuilder/core/bot_adapter.py]: This same function in JS is called runMiddleware and does functionally the same. Which one should I follow?

Comment (cleemullins) [reply libraries/botbuilder-core/botbuilder/core/bot_adapter.py]: I would stick with whichever codebase you've bene (mostly) modeling the Python SDK after. I have no strong preference.

Comment (axelsrz) [reply libraries/botbuilder-core/botbuilder/core/bot_adapter.py]: In previous versions of the Python SDK this function was named after JS. Ultimately I decided to change it given that I'm basing the code for the _most part_ in C# (class names, and structure-wise as well)

File: libraries/botbuilder-core/botbuilder/core/bot_assert.py (hunk 2 above)
Comment (axelsrz) [reply libraries/botbuilder-core/botbuilder/core/bot_assert.py]: Done, actually I just realized that this class was duplicate but this version follows the correct way of none-checking. Removed dup.

File: samples/python-flask/13.core-bot/main.py [hunk 3]
```diff
@@ -29,6 +29,19 @@
 SETTINGS = BotFrameworkAdapterSettings(app.config['APP_ID'], app.config['APP_PASSWORD'])
 ADAPTER = BotFrameworkAdapter(SETTINGS)
 
+# Catch-all for errors.
+async def on_error(context: TurnContext, error: Exception):
+    # This check writes out errors to console log
+    # NOTE: In production environment, you should consider logging this to Azure
+    #       application insights.
+    print(f'\n [on_turn_error]: { error }')
```
Comment (daveta) [inline samples/python-flask/13.core-bot/main.py]: >print(f'\n [on_turn_error]: { error }') [](start = 4, length = 39)

FYI: I've seen where prints get swallowed (can't recall on linux/windows on flask/django).  I've been getting in the practice of using:
print(f"blah-blah", file=sys.stderr)
Which appears to print slightly more consistently.

**Comment by daveta** [comment]
> gi

whoops

---
Refers to: samples/python-flask/13.core-bot/main.py:11 in a078008. [](commit_id = a0780089d5e7dabe752a753a393bb9ea45e0fda0, deletion_comment = False)

File: samples/python-flask/13.core-bot/main.py (hunk 3 above)
Comment (axelsrz) [reply samples/python-flask/13.core-bot/main.py]: solved

**Pull request merged**